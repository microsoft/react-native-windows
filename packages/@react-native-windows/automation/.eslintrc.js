module.exports = {
  extends: ['@rnw-scripts'],
  parserOptions: {tsconfigRootDir : __dirname},
  
  // Workaround "export type" being unsupported with prettier 1.x
  ignorePatterns: ['index.ts'],
  overrides: [
    {
      files: ['*.ts'],
      // Also not clean
      rules: {
        '@typescript-eslint/no-unused-vars': 'off',
      },
    },
  ],
};
