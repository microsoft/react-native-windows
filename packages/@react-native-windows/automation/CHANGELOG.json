{
  "name": "@react-native-windows/automation",
  "entries": [
    {
      "date": "Fri, 27 May 2022 16:26:29 GMT",
      "tag": "@react-native-windows/automation_v0.0.1",
      "version": "0.0.1",
      "comments": {
        "patch": [
          {
            "comment": "Bump @react-native-windows/automation-channel to v0.0.1",
            "author": "vmorozov@microsoft.com",
            "commit": "927b021179ee739610d6ddc81624f30adb4a2d7b",
            "package": "@react-native-windows/automation"
          }
        ]
      }
    },
    {
      "date": "Tue, 24 May 2022 21:08:55 GMT",
      "tag": "@react-native-windows/automation_v0.0.1",
      "version": "0.0.1",
      "comments": {
        "patch": [
          {
            "comment": "Bump @react-native-windows/automation-channel to v0.0.1",
            "author": "julio@rochsquadron.net",
            "commit": "4b1eb4196ab7be2e266da754637a00a252f3745f",
            "package": "@react-native-windows/automation"
          }
        ]
      }
    },
    {
      "date": "Mon, 02 May 2022 15:10:52 GMT",
      "tag": "@react-native-windows/automation_v0.0.1",
      "version": "0.0.1",
      "comments": {
        "patch": [
          {
            "comment": "Bump @react-native-windows/automation-channel to v0.0.1",
            "author": "53799235+ZihanChen-MSFT@users.noreply.github.com",
            "commit": "94d54d33e317332fa9145129d228a8e323f9b948",
            "package": "@react-native-windows/automation"
          }
        ]
      }
    },
    {
      "date": "Mon, 25 Apr 2022 15:09:39 GMT",
      "tag": "@react-native-windows/automation_v0.0.1",
      "version": "0.0.1",
      "comments": {
        "patch": [
          {
            "comment": "Bump @react-native-windows/automation-channel to v0.0.1",
            "author": "vmorozov@microsoft.com",
            "commit": "3e1da9e94cba42c622cf68baecbd57d7e8a6971d",
            "package": "@react-native-windows/automation"
          }
        ]
      }
    },
    {
      "date": "Mon, 04 Apr 2022 15:13:07 GMT",
      "tag": "@react-native-windows/automation_v0.0.1",
      "version": "0.0.1",
      "comments": {
        "patch": [
          {
            "comment": "Bump @react-native-windows/automation-channel to v0.0.1",
            "author": "asklar@microsoft.com",
            "commit": "aab0df703cec0f9e69b185d13a59d976fa62a873",
            "package": "@react-native-windows/automation"
          }
        ]
      }
    },
    {
      "date": "Mon, 14 Mar 2022 15:13:42 GMT",
      "tag": "@react-native-windows/automation_v0.0.1",
      "version": "0.0.1",
      "comments": {
        "patch": [
          {
            "comment": "Bump @react-native-windows/automation-channel to v0.0.1",
            "author": "tudor.mihai@microsoft.com",
            "commit": "7c9bdabed0af6a90c72ea1840efa5b6f00200627",
            "package": "@react-native-windows/automation"
          }
        ]
      }
    },
    {
      "date": "Mon, 28 Feb 2022 16:14:18 GMT",
      "tag": "@react-native-windows/automation_v0.0.1",
      "version": "0.0.1",
      "comments": {
        "patch": [
          {
            "comment": "Bump @react-native-windows/automation-channel to v0.0.1",
            "author": "vmorozov@microsoft.com",
            "commit": "789933a1a3828cdd9cf5ee191689d35b31e134fe",
            "package": "@react-native-windows/automation"
          }
        ]
      }
    },
    {
      "date": "Thu, 24 Feb 2022 16:40:08 GMT",
      "tag": "@react-native-windows/automation_v0.0.1",
      "version": "0.0.1",
      "comments": {
        "patch": [
          {
            "comment": "Bump @react-native-windows/automation-channel to v0.0.1",
            "author": "30809111+acoates-ms@users.noreply.github.com",
            "commit": "9e78137aab868033423df7f4a10a1188f381a3f0",
            "package": "@react-native-windows/automation"
          }
        ]
      }
    },
    {
      "date": "Mon, 21 Feb 2022 16:09:42 GMT",
      "tag": "@react-native-windows/automation_v0.0.1",
      "version": "0.0.1",
      "comments": {
        "patch": [
          {
            "comment": "Bump @react-native-windows/automation-channel to v0.0.1",
            "author": "30809111+acoates-ms@users.noreply.github.com",
            "commit": "7e3d3a9ad44b5ceafceb9aaeb647a0c0193a04c8",
            "package": "@react-native-windows/automation"
          }
        ]
      }
    },
    {
      "date": "Mon, 14 Feb 2022 16:10:47 GMT",
      "tag": "@react-native-windows/automation_v0.0.1",
      "version": "0.0.1",
      "comments": {
        "patch": [
          {
            "comment": "Bump @react-native-windows/automation-channel to v0.0.1",
            "author": "30809111+acoates-ms@users.noreply.github.com",
            "commit": "57f51ad8aaa9ee6759b8935e17095f6821571800",
            "package": "@react-native-windows/automation"
          }
        ]
      }
    },
    {
      "date": "Mon, 07 Feb 2022 16:12:51 GMT",
      "tag": "@react-native-windows/automation_v0.0.1",
      "version": "0.0.1",
      "comments": {
        "patch": [
          {
            "comment": "Bump @react-native-windows/automation-channel to v0.0.1",
            "author": "30809111+acoates-ms@users.noreply.github.com",
            "commit": "9e305803ba81e2375ce6a7cf894a34a23d9002e5",
            "package": "@react-native-windows/automation"
          }
        ]
      }
    },
    {
      "date": "Mon, 31 Jan 2022 16:12:38 GMT",
      "tag": "@react-native-windows/automation_v0.0.1",
      "version": "0.0.1",
      "comments": {
        "patch": [
          {
            "comment": "Bump @react-native-windows/automation-channel to v0.0.1",
            "author": "30809111+acoates-ms@users.noreply.github.com",
            "commit": "0951f3da2e746f1feb0bd7e96d3dd8d5d936d6fb",
            "package": "@react-native-windows/automation"
          }
        ]
      }
    },
    {
      "date": "Mon, 24 Jan 2022 16:11:33 GMT",
      "tag": "@react-native-windows/automation_v0.0.1",
      "version": "0.0.1",
      "comments": {
        "patch": [
          {
            "comment": "Bump @react-native-windows/automation-channel to v0.0.1",
            "author": "ngerlem@microsoft.com",
            "commit": "919d91a3e9845f9a60417fb8e47a36d7caedbc28",
            "package": "@react-native-windows/automation"
          }
        ]
      }
    },
    {
      "date": "Mon, 17 Jan 2022 16:12:35 GMT",
      "tag": "@react-native-windows/automation_v0.0.1",
      "version": "0.0.1",
      "comments": {
        "patch": [
          {
            "comment": "Bump @react-native-windows/automation-channel to v0.0.1",
            "author": "saadnajmi2@gmail.com",
            "commit": "1a70a3b499a0e695597e8669439b856d41096170",
            "package": "@react-native-windows/automation"
          }
        ]
      }
    },
    {
      "date": "Mon, 03 Jan 2022 16:10:55 GMT",
      "tag": "@react-native-windows/automation_v0.0.1",
      "version": "0.0.1",
      "comments": {
        "patch": [
          {
            "comment": "Bump @react-native-windows/automation-channel to v0.0.1",
            "author": "julio.rocha@microsoft.com",
            "commit": "1a0cec7754fdff58dad91fb15c483e884ef70aee",
            "package": "@react-native-windows/automation"
          }
        ]
      }
    },
    {
      "date": "Mon, 22 Nov 2021 16:09:58 GMT",
      "tag": "@react-native-windows/automation_v0.0.1",
      "version": "0.0.1",
      "comments": {
        "patch": [
          {
            "comment": "Bump @react-native-windows/automation-channel to v0.0.1",
            "author": "vmorozov@microsoft.com",
            "commit": "f5e0d1f8184cabdf83784ad272f7b82680b117cf",
            "package": "@react-native-windows/automation"
          }
        ]
      }
    },
    {
      "date": "Mon, 15 Nov 2021 16:09:53 GMT",
      "tag": "@react-native-windows/automation_v0.0.1",
      "version": "0.0.1",
      "comments": {
        "patch": [
          {
            "comment": "Bump @react-native-windows/automation-channel to v0.0.1",
            "author": "julio.rocha@microsoft.com",
            "commit": "9fded4d64856fc7ab1df0040a79466afff3d52fd",
            "package": "@react-native-windows/automation"
          }
        ]
      }
    },
    {
      "date": "Mon, 01 Nov 2021 19:10:27 GMT",
      "tag": "@react-native-windows/automation_v0.0.1",
      "version": "0.0.1",
      "comments": {
        "patch": [
          {
            "comment": "Bump @react-native-windows/automation-channel to v0.0.1",
            "author": "asklar@microsoft.com",
            "commit": "61da34482a055ee52ba23b826d154a7a9fbf0bee",
            "package": "@react-native-windows/automation"
          }
        ]
      }
    },
    {
      "date": "Mon, 25 Oct 2021 15:07:26 GMT",
      "tag": "@react-native-windows/automation_v0.0.1",
      "version": "0.0.1",
      "comments": {
        "patch": [
          {
            "comment": "Bump @react-native-windows/automation-channel to v0.0.1",
            "author": "ngerlem@microsoft.com",
            "commit": "fb9caf3b0384f30a615c4ece8e79f335bbbb4e33",
            "package": "@react-native-windows/automation"
          }
        ]
      }
    },
    {
      "date": "Mon, 18 Oct 2021 15:08:36 GMT",
      "tag": "@react-native-windows/automation_v0.0.1",
      "version": "0.0.1",
      "comments": {
        "patch": [
          {
            "comment": "Bump @react-native-windows/automation-channel to v0.0.1",
            "author": "ngerlem@microsoft.com",
            "commit": "dc11d60eef65114d80fc92e8c7f19c24f3dbacc9",
            "package": "@react-native-windows/automation"
          }
        ]
      }
    },
    {
      "date": "Mon, 11 Oct 2021 15:07:19 GMT",
      "tag": "@react-native-windows/automation_v0.0.1",
      "version": "0.0.1",
      "comments": {
        "patch": [
          {
            "comment": "Bump @react-native-windows/automation-channel to v0.0.1",
            "author": "ngerlem@microsoft.com",
            "commit": "5d1e5ae98139204e02f0ac5d8c06ae7ee1157395",
            "package": "@react-native-windows/automation"
          }
        ]
      }
    },
    {
      "date": "Mon, 04 Oct 2021 15:07:37 GMT",
      "tag": "@react-native-windows/automation_v0.0.1",
      "version": "0.0.1",
      "comments": {
        "prerelease": [
          {
            "comment": "Bump @react-native-windows/automation-channel to v0.0.1",
            "author": "julio.rocha@microsoft.com",
            "commit": "8b1048e4a3491c01f2b2d4382822f8d98f3ccfe5",
            "package": "@react-native-windows/automation"
          }
        ]
      }
    },
    {
      "date": "Mon, 27 Sep 2021 15:07:31 GMT",
      "tag": "@react-native-windows/automation_v0.0.1",
      "version": "0.0.1",
      "comments": {
        "prerelease": [
          {
            "comment": "Bump @react-native-windows/automation-channel to v0.0.1",
            "author": "julio.rocha@microsoft.com",
            "commit": "b4d6793bc39203a7671163f7770307f24f90ef91",
            "package": "@react-native-windows/automation"
          }
        ]
      }
    },
    {
      "date": "Mon, 20 Sep 2021 15:07:01 GMT",
      "tag": "@react-native-windows/automation_v0.0.1",
      "version": "0.0.1",
      "comments": {
        "prerelease": [
          {
            "comment": "Bump @react-native-windows/automation-channel to v0.0.1",
            "author": "34109996+chiaramooney@users.noreply.github.com",
            "commit": "71803345c9e89ee419332119fd5667370c5c19d3",
            "package": "@react-native-windows/automation"
          }
        ]
      }
    },
    {
      "date": "Mon, 13 Sep 2021 15:06:37 GMT",
      "tag": "@react-native-windows/automation_v0.0.1",
      "version": "0.0.1",
      "comments": {
        "prerelease": [
          {
            "comment": "Bump @react-native-windows/automation-channel to v0.0.1",
            "author": "anandrag@microsoft.com",
            "commit": "2afad93679990d06cec763bd6c12ce44efef0c4a",
            "package": "@react-native-windows/automation"
          }
        ]
      }
    },
    {
      "date": "Mon, 06 Sep 2021 15:06:48 GMT",
      "tag": "@react-native-windows/automation_v0.0.1",
      "version": "0.0.1",
      "comments": {
        "prerelease": [
          {
            "comment": "Bump @react-native-windows/automation-channel to v0.0.1",
            "author": "dannyvv@microsoft.com",
            "commit": "8c9852f31fe31e1448b596ed4db4b35f6bec3fd1",
            "package": "@react-native-windows/automation"
          }
        ]
      }
    },
    {
      "date": "Fri, 03 Sep 2021 18:48:19 GMT",
      "tag": "@react-native-windows/automation_v0.0.1",
      "version": "0.0.1",
      "comments": {
        "prerelease": [
          {
            "comment": "Bump @react-native-windows/automation-channel to v0.0.1",
            "author": "ngerlem@microsoft.com",
            "commit": "cf1dcfbe108028869e326bb2bdab370949f7dac8",
            "package": "@react-native-windows/automation"
          }
        ]
      }
    },
    {
      "date": "Fri, 03 Sep 2021 05:06:04 GMT",
      "tag": "@react-native-windows/automation_v0.0.1",
      "version": "0.0.1",
      "comments": {
        "patch": [
          {
            "comment": "@react-native-windows/automation",
            "author": "ngerlem@microsoft.com",
            "commit": "17e32d36180fa5c7e45c3c926c6eaa729efb6ea0",
            "package": "@react-native-windows/automation"
          }
        ],
        "prerelease": [
          {
            "comment": "Bump @react-native-windows/automation-channel to v0.0.1",
            "author": "erozell@outlook.com",
            "commit": "439b8a54b62807068bfa17e130f8ab0378eaaf36",
            "package": "@react-native-windows/automation"
          }
        ]
      }
    }
  ]
}
