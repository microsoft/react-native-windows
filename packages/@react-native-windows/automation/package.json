{
  "name": "@react-native-windows/automation",
  "version": "0.3.11",
  "description": "UI Automation Suite for React Native Windows Applications",
  "main": "lib-commonjs/index.js",
  "repository": "https://github.com/microsoft/react-native-windows",
  "license": "MIT",
  "private": true,
  "scripts": {
    "build": "rnw-scripts build",
    "clean": "rnw-scripts clean",
    "lint": "rnw-scripts lint",
    "lint:fix": "rnw-scripts lint:fix",
    "watch": "rnw-scripts watch"
  },
  "dependencies": {
    "@react-native-windows/automation-channel": "^0.5.11",
    "@react-native-windows/fs": "0.71.0",
    "@typescript-eslint/eslint-plugin": "^5.30.5",
    "@typescript-eslint/parser": "^5.30.5",
    "chalk": "^4.1.2",
    "readline-sync": "1.4.10",
    "webdriverio": "^6.9.0"
  },
  "devDependencies": {
    "@jest/create-cache-key-function": "^29.0.3",
    "@jest/environment": "^29.3.0",
    "@jest/types": "^29.2.1",
    "@rnw-scripts/eslint-config": "1.1.14",
    "@rnw-scripts/just-task": "2.3.3",
    "@rnw-scripts/ts-config": "2.0.2",
    "@types/jest": "^29.2.2",
    "@types/node": "^14.14.22",
    "@types/readline-sync": "^1.4.4",
    "eslint": "^8.19.0",
    "prettier": "^2.4.1",
    "typescript": "^4.9.5"
  },
  "peerDependencies": {
    "jest": ">=29.0.3",
    "jest-environment-node": ">=29.2.2"
  },
  "files": [
    "lib-commonjs",
    "README.md"
  ],
  "engines": {
    "node": ">= 14"
  }
}