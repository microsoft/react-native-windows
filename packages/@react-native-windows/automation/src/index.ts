/**
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * @format
 */

import AutomationEnvironment from './AutomationEnvironment';
import {app, AutomationElement} from './AutomationClient';

export {app};
export type {AutomationElement}
export default AutomationEnvironment;
