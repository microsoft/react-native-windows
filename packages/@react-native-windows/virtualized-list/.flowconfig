[ignore]
; Ignore react-native-windows' build output. Flow is not compiled with long path support and after running unittests these folders have long paths
.*/node_modules/react-native-windows/build/.*
.*/node_modules/react-native-windows/target/.*
.*/node_modules/react-native-windows/Microsoft.ReactNative.Managed.CodeGen.UnitTests/.*
.*/node_modules/react-native-windows/Microsoft.ReactNative.Managed.UnitTests/.*

[libs]
../../../node_modules/react-native/interface.js
../../../node_modules/react-native/flow/

[include]
; Need to include hoisted modules
../../../node_modules/

[options]
emoji=true

exact_by_default=true

format.bracket_spacing=false

module.file_ext=.js
module.file_ext=.json
module.file_ext=.win32.js

munge_underscores=true

suppress_type=$FlowIssue
suppress_type=$FlowFixMe
suppress_type=$FlowFixMeProps
suppress_type=$FlowFixMeState
suppress_type=$FlowFixMeEmpty

[lints]
sketchy-null-number=warn
sketchy-null-mixed=warn
sketchy-number=warn
untyped-type-import=warn
nonstrict-import=warn
deprecated-type=warn
unsafe-getters-setters=warn
unnecessary-invariant=warn
signature-verification-failure=warn

[strict]
deprecated-type
nonstrict-import
sketchy-null
unclear-type
unsafe-getters-setters
untyped-import
untyped-type-import

[version]
^0.152.0
