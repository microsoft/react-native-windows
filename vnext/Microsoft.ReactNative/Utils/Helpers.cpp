// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

#include "pch.h"

#include <Modules/NativeUIManager.h>
#include <UI.Xaml.Media.h>
#include <Utils/Helpers.h>
#include <winrt/Windows.Foundation.Metadata.h>

#include <Modules/PaperUIManagerModule.h>
#include <appmodel.h>
#include <processthreadsapi.h>

#ifdef USE_FABRIC
#include <Fabric/Composition/CompositionUIService.h>
#endif // USE_FABRIC

namespace winrt {
using namespace xaml::Controls::Primitives;
using namespace xaml::Media;
using namespace Windows::Foundation::Metadata;
} // namespace winrt

namespace Microsoft::ReactNative {

// Not only react-native, native modules could set tag too for controls.
// For example, to identify an clicked item, customer may add tag in
// NavigationView since content for the two NavigationViewItem are empty.
//
// <NavigationView>
//  <NavigationViewItem Icon="Accept" Tag="1" />
//  <NavigationViewItem Icon="Accept" Tag="2" />
// </NavigationView>
// Instead of deduce view id directly from FrameworkElement.Tag, this do
// additional check by uimanager.
ReactId getViewId(const Mso::React::IReactContext &context, xaml::FrameworkElement const &fe) {
  ReactId reactId{};
  if (auto uiManager = Microsoft::ReactNative::GetNativeUIManager(context).lock()) {
    if (auto peer = uiManager->reactPeerOrContainerFrom(fe)) {
      reactId.isValid = true;
      reactId.tag = Microsoft::ReactNative::GetTag(peer);
    }
  }
  return reactId;
};

std::int32_t CountOpenPopups() {
  // TODO: Use VisualTreeHelper::GetOpenPopupsFromXamlRoot when running against
  // RS6
  winrt::Windows::Foundation::Collections::IVectorView<xaml::Controls::Primitives::Popup> popups =
      xaml::Media::VisualTreeHelper::GetOpenPopups(xaml::Window::Current());
  return (int32_t)popups.Size();
}

template <uint16_t APIVersion>
bool IsAPIContractVxAvailable() {
  static bool isAPIContractVxAvailableInitialized = false;
  static bool isAPIContractVxAvailable = false;
  if (!isAPIContractVxAvailableInitialized) {
    isAPIContractVxAvailableInitialized = true;
    isAPIContractVxAvailable =
        winrt::ApiInformation::IsApiContractPresent(L"Windows.Foundation.UniversalApiContract", APIVersion);
  }

  return isAPIContractVxAvailable;
}

bool IsAPIContractV5Available() {
  return IsAPIContractVxAvailable<5>();
}

bool IsAPIContractV6Available() {
  return IsAPIContractVxAvailable<6>();
}

bool IsAPIContractV7Available() {
  return IsAPIContractVxAvailable<7>();
}

bool IsAPIContractV8Available() {
  return IsAPIContractVxAvailable<8>();
}

bool IsAPIContractV12Available() {
  return IsAPIContractVxAvailable<12>();
}

bool IsRS3OrHigher() {
  return IsAPIContractV5Available();
}

bool IsRS4OrHigher() {
  return IsAPIContractV6Available();
}

bool IsRS5OrHigher() {
  return IsAPIContractV7Available();
}

bool Is19H1OrHigher() {
  return IsAPIContractV8Available();
}

bool Is21H1OrHigher() {
  return IsAPIContractV12Available();
}

bool IsXamlIsland() {
  AppPolicyWindowingModel e;
  if (FAILED(AppPolicyGetWindowingModel(GetCurrentThreadEffectiveToken(), &e)) ||
      e == AppPolicyWindowingModel_ClassicDesktop) {
    return true;
  }
  return false;
}

bool IsWinUI3Island() {
#ifndef USE_WINUI3
  return false;
#else
  return IsXamlIsland();
#endif
}

bool IsFabricEnabled(winrt::Microsoft::ReactNative::IReactPropertyBag const &properties) {
#ifdef USE_FABRIC
  return winrt::Microsoft::ReactNative::Composition::implementation::CompositionUIService::GetCompositionContext(
             properties) != nullptr;
#else
  return false;
#endif
}

} // namespace Microsoft::ReactNative
