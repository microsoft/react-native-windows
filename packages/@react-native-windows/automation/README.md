# @react-native-windows/automation

`@react-native-windows/automation` makes it easy to perform UI tests against
your `react-native-windows` application using Jest. 

**This package is a work in progress**

## System Requirements

`@react-native-windows/automation` relies on [WinAppDriver 1.2.1](https://github.com/microsoft/WinAppDriver/releases/tag/v1.2.1)
or later to manipulate and inspect your application via [Windows UI Automation](https://docs.microsoft.com/en-us/dotnet/framework/ui-automation/ui-automation-overview).
This is already installed in common CI environments like GitHub Actions or
Azure Pipelines hosted images.

## Configuring Jest

First, ensure that `@react-native-windows/automation` is included as a
dependency in your package:

```json
{
  "devDependencies": {
    "@react-native-windows/automation": "<version>",
  }
}
```

Next, edit your [Jest configuration](https://jestjs.io/docs/configuration) to
use `@react-native-windows/automation` as your test environment. Add the
following to your `jest.config.js` or `jest.config.ts` file:

```js
module.exports = {
...
  // Set up the automation environment before running tests
  testEnvironment: '@react-native-windows/automation',

  // Only run a single test at a time
  maxWorkers: 1,

  // Set up @react-native-windows/automation specific options (see below)
  testEnvironmentOptions: {
    app: '<AppName>',
  },
};
```

## Environment Options

`@react-native-windows/automation` is configured via the
`testEnvironmentOptions` property in your Jest configuration. The following
options are valid:

```js
testEnvironmentOptions: {
  // Required: Your application to launch, as either an AppX package identity,
  // or path to an .exe for an unpackaged application.
  app: 'Microsoft.WindowsAlarms',

  // Optional: Explicit path to WinAppDriver. By default,
  // `@react-native-windows/automation` tries to use
  // "%PROGRAMFILES(X86)%\Windows Application Driver\WinAppDriver.exe"
  winAppDriverBin: 'D:\\WinAppDriver.exe',

  // Optional: Options to be passed to WebDriverIO
  // See https://webdriver.io/docs/options/
  webdriverOptions: {
      // Port to use for WebDriver server (Default 4723)
      port: 4444

      // Level of logging verbosity: trace | debug | info | warn | error
      logLevel: 'error',

      // Default timeout for all waitFor* commands.
      waitforTimeout: 60000,

      // Default timeout in milliseconds for request
      connectionRetryTimeout: 90000,

      // Default request retries count
      connectionRetryCount: 10,
    },
},
```

## Writing Tests

`@react-native-windows/automation` exports an `app` member with functions to
perform globally scoped WebDriver commands, such as locating an element or
waiting for a condition. Several strategies are present to locate an element,
with the reccomended being to use React Native's `testID` property.

A located `AutomationElement` exposes methods to allow application interaction
and introspection, such as clicking or typing.

```js
import {app} from '@react-native-windows/automation';

beforeAll(async () => {
  const appHeader = await app.findElementByTestID('app-header');

  await app.waitUntil(async () => {
    const headerText = await appHeader.getText();
    return headerText.includes('Welcome');
  }) 
});

test('Type abc', async () => {
  const textInput = await app.findElementByTestID('textinput-field');
  await textInput.setValue('abc');
  expect(await textInput.getText()).toBe('abc');
});
```

## Performing Additional Actions

**WIP Not yet exposed**

Sometimes it is useful to exmaine or manipulate the application in ways that
are not exposed to Windows UI Automation. An additional package,
`@react-native-windows/automation-channel` can be included in your application
as a native module to allow writing custom commands performed by your
application in-process. Some pre-built commands can be included using the
`@react-native-windows/automation-commands` package.

## Visual Comparison

**WIP Not yet exposed**

A common use-case of UI testing is to ensure your components look the way you
expect. Tools like [enzyme](https://github.com/enzymejs/enzyme) allow
inspecting your React component tree, but are oblivious to the native UI tree.

`@react-native-windows/automation-commands` exposes a `dumpVisualTree` command
which creates a JSON object corresponding to the XAML tree under a given
testID. This can be used in conjunction with snapshot testing to validate
native component logic stays consistent.

```js
import {dumpVisualTree} from '@react-native-windows/automation-commands';

test('Widget', async () => {
  const dump = await dumpVisualTree('widget-test-id');
  expect(dump).toMatchSnapshot();
});
```