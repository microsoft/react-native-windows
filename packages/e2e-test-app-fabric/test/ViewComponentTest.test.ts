/**
 * Copyright (c) Microsoft Corporation.
 * Licensed under the MIT License.
 *
 * @format
 */

//import {dumpVisualTree} from '@react-native-windows/automation-commands';
import {goToComponentExample} from './RNTesterNavigation';
import {app} from '@react-native-windows/automation';
import {verifyNoErrorLogs} from './Helpers';

beforeAll(async () => {
  // If window is partially offscreen, tests will fail to click on certain elements
  await app.setWindowPosition(0, 0);
  await app.setWindowSize(1000, 1250);
  await goToComponentExample('View');
});

afterEach(async () => {
  await verifyNoErrorLogs();
});

describe('View Tests', () => {
  test('Control', () => {
    expect(true).toBe(true);
  });
  // Disable until tests pass in CI.
  /*test('Views can have border styles', async () => {
    const searchBox = await app.findElementByTestID('example_search');
    await searchBox.addValue(['Backspace', 'Backspace', 'Backspace']);
    await searchBox.addValue('sty');
    const componentsTab = await app.findElementByTestID('border-style-button');
    await componentsTab.waitForDisplayed({timeout: 5000});
    const dump = await dumpVisualTree('border-style-button');
    expect(dump).toMatchSnapshot();
  });
  test('Views can have offscreen alpha compositing', async () => {
    const searchBox = await app.findElementByTestID('example_search');
    await searchBox.addValue(['Backspace', 'Backspace', 'Backspace']);
    await searchBox.addValue('off');
    const componentsTab = await app.findElementByTestID('offscreen-alpha-compositing-button');
    await componentsTab.waitForDisplayed({timeout: 5000});
    const dump = await dumpVisualTree('offscreen-alpha-compositing-button');
    expect(dump).toMatchSnapshot();
  });
  test('Views can have a z-index', async () => {
    const searchBox = await app.findElementByTestID('example_search');
    await searchBox.addValue(['Backspace', 'Backspace', 'Backspace']);
    await searchBox.addValue('z');
    const componentsTab = await app.findElementByTestID('z-index-button');
    await componentsTab.waitForDisplayed({timeout: 5000});
    const dump = await dumpVisualTree('z-index-button');
    expect(dump).toMatchSnapshot();
  });
  test('Views can have display: none', async () => {
    const searchBox = await app.findElementByTestID('example_search');
    await searchBox.addValue(['Backspace', 'Backspace', 'Backspace']);
    await searchBox.addValue('dis');
    const componentsTab = await app.findElementByTestID('display-none-button');
    await componentsTab.waitForDisplayed({timeout: 5000});
    const dump = await dumpVisualTree('display-none-button');
    expect(dump).toMatchSnapshot();
  });
  */
});
