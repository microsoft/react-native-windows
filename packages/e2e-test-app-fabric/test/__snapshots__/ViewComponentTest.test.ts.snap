// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View Tests Views can have background color 1`] = `
{
  "Automation Tree": {
    "AutomationId": "background-color",
    "Children": [
      {
        "AutomationId": "",
        "ControlType": 50020,
        "HelpText": "",
        "IsEnabled": true,
        "IsKeyboardFocusable": false,
        "LocalizedControlType": "text",
        "Name": "Blue background",
      },
    ],
    "ControlType": 50026,
    "HelpText": "",
    "IsEnabled": true,
    "IsKeyboardFocusable": false,
    "LocalizedControlType": "group",
    "Name": "",
  },
  "Visual Tree": {
    "Brush": {
      "Brush Type": "ColorBrush",
      "Color": "rgba(0, 128, 0, 255)",
    },
    "Children": [
      {
        "Children": [
          {
            "Opacity": 1,
            "Size": [
              1020,
              16,
            ],
            "Visual Type": "SpriteVisual",
          },
          {
            "Opacity": 1,
            "Size": [
              0,
              0,
            ],
            "Visual Type": "SpriteVisual",
          },
        ],
        "Opacity": 1,
        "Size": [
          1020,
          16,
        ],
        "Visual Type": "SpriteVisual",
      },
    ],
    "Comment": "background-color",
    "Opacity": 1,
    "Size": [
      1030,
      24,
    ],
    "Visual Type": "SpriteVisual",
  },
}
`;

exports[`View Tests Views can have opacity 1`] = `
{
  "Automation Tree": {
    "AutomationId": "opacity-1",
    "Children": [
      {
        "AutomationId": "",
        "ControlType": 50020,
        "HelpText": "",
        "IsEnabled": true,
        "IsKeyboardFocusable": false,
        "LocalizedControlType": "text",
        "Name": "Opacity 0.1",
      },
    ],
    "ControlType": 50026,
    "HelpText": "",
    "IsEnabled": true,
    "IsKeyboardFocusable": false,
    "LocalizedControlType": "group",
    "Name": "",
  },
  "Visual Tree": {
    "Children": [
      {
        "Children": [
          {
            "Opacity": 1,
            "Size": [
              1030,
              19,
            ],
            "Visual Type": "SpriteVisual",
          },
          {
            "Opacity": 1,
            "Size": [
              0,
              0,
            ],
            "Visual Type": "SpriteVisual",
          },
        ],
        "Opacity": 1,
        "Size": [
          1030,
          19,
        ],
        "Visual Type": "SpriteVisual",
      },
    ],
    "Comment": "opacity-1",
    "Opacity": 0.10000000149011612,
    "Size": [
      1030,
      19,
    ],
    "Visual Type": "SpriteVisual",
  },
}
`;
